src/hal/src/pso/descriptor.rs

```diff
@@ -119,23 +119,32 @@ pub trait DescriptorPool<B: Backend>: Send + Sync + fmt::Debug {
     /// Descriptors will become invalid once the pool is reset. Usage of invalidated descriptor sets results
     /// in undefined behavior.
     fn allocate_set(&mut self, layout: &B::DescriptorSetLayout) -> Result<B::DescriptorSet, AllocationError> {
-        self.allocate_sets(Some(layout)).remove(0)
+        let mut sets = Vec::with_capacity(1);
+        self.allocate_sets(Some(layout), &mut sets)
+            .map(|_| sets.remove(0))
     }
 
     /// Allocate one or multiple descriptor sets from the pool.
     ///
     /// Each descriptor set will be allocated from the pool according to the corresponding set layout.
     /// Descriptors will become invalid once the pool is reset. Usage of invalidated descriptor sets results
     /// in undefined behavior.
-    fn allocate_sets<I>(&mut self, layouts: I) -> Vec<Result<B::DescriptorSet, AllocationError>>
+    fn allocate_sets<I>(&mut self, layouts: I, sets: &mut Vec<B::DescriptorSet>) -> Result<(), AllocationError>
```

Comment:
I took a few attempts to make this API nicer before, and failed. `impl Iterator` still has to be one concrete type, and different backends would not agree on it. Having it as an associated type of this trait then prevents any sort of default implementation... You are welcome to try it out :)

Reply:
It would be cool to figure this out, because I think we could use that pattern in a few places. Anyway I don't think it's worth holding up this PR, we can change it later if we figure it out (or if `impl Iterator` widens type support)